### ISSUE TITLE
Type guard is broken if the block contains assignment to the guarded variable

### ISSUE THREAD

**Issue by cervengoc** [Microsoft/TypeScript#8022]
**Expected behavior**

This function should compile succesfully
```
function testFunction(a: string | Object) {
  if (typeof a === "string") {
    if (a.indexOf(".") >= 0)
      a = a.substr(a.indexOf(".") + 1);
    ...
  }
}
```

**Actual behavior**

The function gives error about `indexOf` does not exist on type `string | Object`. If I omit the assignment to `a` then it works fine. I think the compiler should notice that I'm safely assigning another `string` to the `string` type-guarded variable. However, I understand that this behavior would be expected if I assigned a value which is not guaranteed to be a `string`.

**TypeScript environment**

Visual Studio 2015
TypeScript 1.8.4

**Comment by RyanCavanaugh** [comment]
This is being addressed by #8010 and has been filed a bunch of times already

**Comment by cervengoc** [comment]
Thank you, sorry for the duplicate. I wasn't searching for the correct keywords.
